### ISSUE TITLE
Add capacity to run in time-backward mode

### ISSUE THREAD

**Issue by erikvansebille** [OceanPARCELS/parcels#12]
For the GMD manuscript, we will need to have functionality to run particles in time-backward mode. As for the integration, this can be done by simply negating velocity fields. The trick is to also handle input fields in reverse order

**Comment by erikvansebille** [comment]
OK, so I'm started looking into this. Initially, I thought I would just write a simple kernel 
```
def reverseUV(particle, grid, time, dt):
    grid.U[time, particle.lon, particle.lat] = -1 * grid.U[time, particle.lon, particle.lat]
    grid.V[time, particle.lon, particle.lat] = -1 * grid.V[time, particle.lon, particle.lat]
```
And then prepend that kernel to the `pset.execute` set

However, this will not work for three reasons:

1. While the right hand sides of these equations can be executes, they can't be assigned to the left hand side because `[time, particle.lon, particle.lat]` are floats, rather than integers
2. Even if we would be able to make this work, it will only negate velocities on the location where the particles are now. However, RK4 will also require the velocities at the intermediate points to be negated
3. There is no way to check that we haven't doubly-negated a velocity field if we call reverseUV for the same location again

So really, it seems that we have a few options: 

1. Put the negation inside the Advection* kernels (with the disadvantage that is a user defines its own kernel it will not necessarily use the same negated velocities)
2. Negate all the grid cells around the particles (although this will not fix problem 3. above)
3. Negate the entire velocity field when we read it in

Option 3. is the easiest to implement, but it can also be quite slow. In scipy mode python will probably do some under the hood vector magic; but not sure how this works in JIT mode. @mlange05 any thoughts?

**Comment by mlange05** [comment]
Ok, I hadn't really thought about this problem in detail yet. Option 3 should work for both modes, but will probably be costly, as you point out. So how about option 4: Optionally perform negation in-place as part of the sampling routine. What I had in mind is this:
* Add an attribute `negate` to `Field` objects and ensure that it is `True` for U and V if we call `execute(..., reverse=True)`. We don't want to negate temperature, right?
* For Python, we then adjust `Field.eval()` to do:
```
val = ...
if self.negate:
    val *= -1.
return val
```
* For JIT mode we add a function `temporal_interpolation_linear_reverse`, which directly calls the original interpolation function `temporal_interpolation_linear`, but negates the return value. Then all we need to do is conditionally replace the function name in `Field.ccode_subscript` with the reverse function and we should achieve the desired in-place auto-negation. Does this sound sensible?

**Comment by erikvansebille** [comment]
OK, so I've almost implemented time-backward mode. The branch now works now for scipy, and in principle also for JIT. However, test_moving_eddies doesn’t work yet because something strange is happening during compilation. 

If I run only in backward mode both JIT and scipy work fine, but if I do a test where I first advect particles forward in time and then reverse velocity files to advect them backward in time, the particles should end up at the same location where they started (within to RK4 errors). 

This works perfectly in scipy mode (see the `moving_eddies_fwdbwd` function). But in JIT mode, the velocity fields do not get negated for some reason. Even if we recompile the code after reversing time (and the compiled .c file does call the `_negate` function), the particles are still advected with non-negated fields. It's as if the new, second .c file is not used..

Not sure what's going on. @mlange05 any ideas?

**Comment by mlange05** [comment]
Ok, I don't think removing the (still very minimal) kernel compilation caching we do is the right way to go (see inline comment on ef16c88). Instead we probably want to avoid the need for recompilation here altogether. 

So, how about this: Instead of calling a different interpolation function, we could store a `float dtsign` on the `struct CField` in `parcels.h` and pass this down as either `1` or `-1` when creating the struct in `Field.ctypes_struct`. All we need to do then is multiply the result of `temporal_interpolation_linear` with `dtsign` in-place and set the field according to `Field.negate`. Does this sound like better solution?

**Comment by erikvansebille** [comment]
OK, I see your point on removing the kernel compilation caching. I like your idea of adding a `float dtsign`, will implement that when I have time!

**Comment by erikvansebille** [comment]
Implemented in PR #52
